contracts/strategies/morpho/MetaMorphoStrategy.sol

```diff
@@ -149,11 +203,194 @@ contract MetaMorphoStrategy is ERC4626Strategy {
+      apr = _rewardsInUnderlyings * _vaultShare * 100 / _totalSupplyAssets;
+    }
+  }
+
+  /// @notice quote rewards per year to underlyings
+  /// @param _rewardData reward data
+  function _quoteRewards(RewardData memory _rewardData) internal view returns (uint256) {
+    uint256 _oneReward = 10 ** uint256(IERC20Detailed(_rewardData.rewardToken).decimals());
+    // quote 1 reward to underlyings then multiply by supplyRewardTokensPerYear
+    return IStaticQuoter(UNI_V3_STATIC_QUOTER).quoteExactInput(_rewardData.uniV3Path, _oneReward) * 
+      IRewardsEmissions(MORPHO_REWARDS_EMISSION).rewardsEmissions(
+        _rewardData.sender,
+        _rewardData.urd,
+        _rewardData.rewardToken, 
+        _rewardData.marketId
+      ).supplyRewardTokensPerYear / _oneReward;
+  }
+
+  /// @notice calculate how much of vault `_add` amount will be added to this market
+  /// @param _mmVault metamorpho vault
+  /// @param _targetMarketId target market id
+  /// @param _supplyQueueLen supply queue length
+  /// @param _add amount of liquidity to add
+  function _calcMarketAdd(
+    IMMVault _mmVault,
+    bytes32 _targetMarketId,
+    uint256 _supplyQueueLen,
+    uint256 _add
+  ) internal view returns (uint256) {
+    IMorpho.Market memory _market;
+    // loop throuh supplyQueue, starting from the first market, and see how much will
+    // be deposited in target market
+    for (uint256 i = 0; i < _supplyQueueLen; i++) {
+      bytes32 _currMarketId = _mmVault.supplyQueue(i);
+      _market = MORPHO_BLUE.market(_currMarketId);
+      // get max depositable amount for this market
+      (uint184 _marketCap,,) = _mmVault.config(_targetMarketId);
+      uint256 _maxDeposit = uint256(_marketCap) - _market.totalSupplyAssets;
+      // If this is the target market, return the current _add value, eventually
+      // reduced to the max depositable amount
+      if (_currMarketId == _targetMarketId) {
+        if (_add > _maxDeposit) {
+          _add = _maxDeposit;
+        }
+        break;
+      }
+      // If this is not the target market, check if we can deposit all the _add amount
+      // in this market, otherwise continue the loop and subtract the max depositable
+      if (_add > _maxDeposit) {
+        _add -= _maxDeposit;
+      } else {
+        _add = 0;
+        break;
+      }
+    }
+
+    return _add;
+  }
+
+  /// @notice calculate how much of vault `_sub` amount will be removed from target market
+  /// @param _mmVault metamorpho vault
+  /// @param _targetMarketId target market id
+  /// @param _withdrawQueueLen withdraw queue length
+  /// @param _sub liquidity to remove
+  function _calcMarketSub(
+    IMMVault _mmVault, 
+    bytes32 _targetMarketId,
+    uint256 _withdrawQueueLen,
+    uint256 _sub
+  ) internal view returns (uint256) {
+    IMorpho.Market memory _market;
+    // loop throuh withdrawQueue, and see how much will be redeemed in target market
+    for (uint256 i = 0; i < _withdrawQueueLen; i++) {
```

Comment:
I this second loop on _withdrawQueueLen can be avoided (to not have a N^2 complexity).
You could remove the loop in this function and pass `_aprData.sub` to `_calcMarketSub` and reduce it by the amount that can be withdrawn from the market.

Reply:
I'm not sure I'm getting what you mean here can you expand on this? O(N^2) seems unavoidable as in theory I have to check all markets until I get to target market because there could be markets before the target market (eg 'idle' market) where I can withdraw all the liquidity

Reply:
adding ref for this https://github.com/Idle-Labs/idle-tranches/pull/86